src/referenced.cc

```diff
@@ -0,0 +1,171 @@
+        return false;
+      }
+      name = reference.words(idx);
+    }
+
+    if (match.condition() == ReferencedAttributes::ABSENCE) {
+      absence_keys_.push_back(name);
+    } else if (match.condition() == ReferencedAttributes::EXACT) {
+      exact_keys_.push_back(name);
+    } else if (match.condition() == ReferencedAttributes::REGEX) {
+      // Don't support REGEX yet, return false to no caching the response.
+      GOOGLE_LOG(ERROR) << "Received REGEX in ReferencedAttributes for "
+                        << name;
+      return false;
+    }
+  }
+
+  return true;
+}
+
+bool Referenced::Signature(const Attributes& attributes,
+                           std::string* signature) const {
+  for (const std::string& key : absence_keys_) {
+    // if an "absence" key exists, return false for mis-match.
+    if (attributes.attributes.find(key) != attributes.attributes.end()) {
+      return false;
+    }
+  }
+
+  MD5 hasher;
+  for (const std::string& key : exact_keys_) {
+    const auto it = attributes.attributes.find(key);
+    // if an "exact" attribute not present, return false for mismatch.
+    if (it == attributes.attributes.end()) {
+      return false;
+    }
+
+    hasher.Update(it->first);
+    hasher.Update(kDelimiter, kDelimiterLength);
+    switch (it->second.type) {
+      case Attributes::Value::ValueType::STRING:
+        hasher.Update(it->second.str_v);
+        break;
+      case Attributes::Value::ValueType::BYTES:
+        hasher.Update(it->second.str_v);
+        break;
+      case Attributes::Value::ValueType::INT64:
+        hasher.Update(&it->second.value.int64_v,
+                      sizeof(it->second.value.int64_v));
+        break;
+      case Attributes::Value::ValueType::DOUBLE:
+        hasher.Update(&it->second.value.double_v,
+                      sizeof(it->second.value.double_v));
+        break;
+      case Attributes::Value::ValueType::BOOL:
+        hasher.Update(&it->second.value.bool_v,
+                      sizeof(it->second.value.bool_v));
+        break;
+      case Attributes::Value::ValueType::TIME:
+        hasher.Update(&it->second.time_v, sizeof(it->second.time_v));
+        break;
+      case Attributes::Value::ValueType::DURATION:
+        hasher.Update(&it->second.duration_nanos_v,
+                      sizeof(it->second.duration_nanos_v));
+        break;
+      case Attributes::Value::ValueType::STRING_MAP:
+        for (const auto& sub_it : it->second.string_map_v) {
+          hasher.Update(sub_it.first);
+          hasher.Update(kDelimiter, kDelimiterLength);
+          hasher.Update(sub_it.second);
+          hasher.Update(kDelimiter, kDelimiterLength);
+        }
+        break;
+    }
+    hasher.Update(kDelimiter, kDelimiterLength);
+  }
+
+  *signature = hasher.Digest();
+  return true;
+}
+
+std::string Referenced::Hash() const {
+  MD5 hasher;
+  // Use std::set to sort the words
+  std::set<std::string> sorted_absence_keys(absence_keys_.begin(),
```

Comment:
Since `Hash()` is relatively called frequently, why don't we just keep `absense_keys` and `exact_keys` always sorted? Having a sorted set calculated every time looks inefficient.

Reply:
For each Check response,  Hash() is called once.  Only Hash() need to sort.  Other usage are just sequential use, not need to sort.